typing/ctype.ml

```diff
@@ -689,8 +689,16 @@ let check_scope_escape level ty =
       );
       begin match ty.desc with
       | Tconstr (p, _, _) when level < Path.scope p ->
-        let var = newvar2 level in
-        raise (Unify [(ty,ty); (var, var)])
+          begin try
+            let ty' = !forward_try_expand_once env ty in
+            aux ty'
```

Comment:
I don't think a non-tail-rec call to `aux` is the right thing to do here. `match !forward... with exception Cannot_expand` would make it tail-recursive again, but anyway this doesn't work well with `iter_type_expr aux ty` that follows.

One thing you could do is

```ocaml
let rec aux ty =
  ...
and continue ty = iter_type_expr aux ty
```

and call `continue` in each branch of the `ty.desc` matching.

Reply:
I’d be happy to switch to `match with exception` (I think it’s better style, I had just simply copied what I found in `update_level`) here, but I’m not sure it’s worth going further than that, especially considering the proposition is not really nicer / clearer.
Again, my current best guess is that this branch (the non-exceptional one) is dead code. And even if it wasn’t, you’d need a long chain of equalities to get that piece of code to stackoverflow.

Reply:
But the traversal is really wrong here, you recurse on the new type (with this call to `aux`) *and* on the old type (with the following call to `iter_type_expr`). This means that you could have exponential behaviors on deeply nested parametrized types like `'a list list list list list list list list`. You need to make sure that `iter_type_expr` is only called in the *other* cases than `Tconstr`, which is why I suggested the `continue` pattern to make it concise to call it several times.

Reply:
Ah! Yes indeed :)
Thanks for the review, I amended my last commit based on your remarks.